include/nanobind/stl/chrono.h

```diff
@@ -0,0 +1,439 @@
+        if (unpack_timedelta(src.ptr(), &dd, &ss, &uu)) {
+            value = type(duration_cast<duration<rep, period>>(
+                             days(dd) + seconds(ss) + microseconds(uu)));
+            return true;
+        } else if (PyErr_Occurred()) {
+            return false;
+        }
+
+        // If invoked with a float we assume it is seconds and convert
+        int is_float;
+#ifdef Py_LIMITED_API
+        is_float = PyObject_IsInstance(src.ptr(), (PyObject *) &PyFloat_Type);
+        if (is_float < 0) {
+            return false;
+        }
+#else
+        is_float = PyFloat_Check(src.ptr());
+#endif
+        if (is_float) {
+            value = type(duration_cast<duration<rep, period>>(duration<double>(PyFloat_AsDouble(src.ptr()))));
+            return true;
+        }
+        return false;
+    }
+
+    // If this is a duration just return it back
+    static const std::chrono::duration<rep, period>& get_duration(const std::chrono::duration<rep, period> &src) {
+        return src;
+    }
+
+    // If this is a time_point get the time_since_epoch
+    template <typename Clock> static std::chrono::duration<rep, period> get_duration(const std::chrono::time_point<Clock, std::chrono::duration<rep, period>> &src) {
+        return src.time_since_epoch();
+    }
+
+    static handle from_cpp(const type &src, rv_policy, cleanup_list*) noexcept {
+        using namespace std::chrono;
+
+        // Use overloaded function to get our duration from our source
+        // Works out if it is a duration or time_point and get the duration
+        auto d = get_duration(src);
+
+        // Declare these special duration types so the conversions happen with the correct primitive types (int)
+        using dd_t = duration<int, std::ratio<86400>>;
+        using ss_t = duration<int, std::ratio<1>>;
+        using us_t = duration<int, std::micro>;
+
+        auto dd = duration_cast<dd_t>(d);
+        auto subd = d - dd;
+        auto ss = duration_cast<ss_t>(subd);
+        auto us = duration_cast<us_t>(subd - ss);
+        return pack_timedelta(dd.count(), ss.count(), us.count());
+    }
+
+    NB_TYPE_CASTER(type, const_name("datetime.timedelta"));
+};
+
+template <class... Args>
+auto can_localtime_s(Args*... args) ->
+    decltype((localtime_s(args...), std::true_type{}));
+std::false_type can_localtime_s(...);
+
+template <class... Args>
+auto can_localtime_r(Args*... args) ->
+    decltype((localtime_r(args...), std::true_type{}));
+std::false_type can_localtime_r(...);
+
+template <class Time, class Buf>
+inline std::tm *localtime_thread_safe(const Time *time, Buf *buf) {
+    if constexpr (decltype(can_localtime_s(time, buf))::value) {
+        // C11 localtime_s
+        std::tm* ret = localtime_s(time, buf);
+        return ret;
+    } else if constexpr (decltype(can_localtime_s(buf, time))::value) {
+        // Microsoft localtime_s (with parameters switched and errno_t return)
+        int ret = localtime_s(buf, time);
```

Comment:
Ugh, what a mess. I did not realize that there were platform differences in such a core API.

Reply:
Yep; the pybind caster tried one form of localtime_s and then fell back to localtime + std::mutex if that wasn't available.